File: frontend/src/domain/pace.ts
import type { Checkpoint, CheckpointLog, Runner } from './types'

export interface RunnerPace {
  paceMinPerMile: number | null
  lastLoggedDist: number | null
  lastLoggedAt: Date | null
}

interface LoggedCheckpoint {
  log: CheckpointLog
  cp: Checkpoint
}

// A DNS/DNF/MOVED/FINISHED runner has no meaningful pace or splits.
function hasNoMeaningfulPace(runner: Runner): boolean {
  return (
    runner.Status === 'DNS' ||
    runner.Status === 'DNF' ||
    runner.Status === 'MOVED' ||
    runner.Status === 'FINISHED'
  )
}

// This runner's logged checkpoints that have a known distance, sorted by display order.
function loggedWithDistance(
  runnerID: number,
  checkpoints: Checkpoint[],
  logs: CheckpointLog[],
): LoggedCheckpoint[] {
  const cpMap = new Map(checkpoints.map((cp) => [cp.ID, cp]))
  return logs
    .filter((l) => l.RunnerID === runnerID)
    .map((l) => ({ log: l, cp: cpMap.get(l.CheckpointID) }))
    .filter((x): x is LoggedCheckpoint => x.cp !== undefined && x.cp.DistanceFromStart !== null)
    .sort((a, b) => a.cp.DisplayOrder - b.cp.DisplayOrder)
}

// Pace between two logged-with-distance checkpoints, or null if the delta is zero/negative.
function pairPace(prev: LoggedCheckpoint, cur: LoggedCheckpoint): number | null {
  const distDelta = cur.cp.DistanceFromStart! - prev.cp.DistanceFromStart!
  const timeDeltaMs =
    new Date(cur.log.RecordedAt).getTime() - new Date(prev.log.RecordedAt).getTime()
  return distDelta > 0 && timeDeltaMs > 0 ? timeDeltaMs / 60000 / distDelta : null
}

// Computes pace from the last two logged checkpoints that both have a known distance.
// DNS/DNF/MOVED/FINISHED runners return nulls.
export function computeRunnerPace(
  runner: Runner,
  checkpoints: Checkpoint[],
  logs: CheckpointLog[],
): RunnerPace {
  if (hasNoMeaningfulPace(runner)) {
    return { paceMinPerMile: null, lastLoggedDist: null, lastLoggedAt: null }
  }

  const withDist = loggedWithDistance(runner.ID, checkpoints, logs)

  if (withDist.length < 2) {
    const last = withDist[0] ?? null
    return {
      paceMinPerMile: null,
      lastLoggedDist: last?.cp.DistanceFromStart ?? null,
      lastLoggedAt: last ? new Date(last.log.RecordedAt) : null,
    }
  }

  const prev = withDist[withDist.length - 2]
  const last = withDist[withDist.length - 1]

  return {
    paceMinPerMile: pairPace(prev, last),
    lastLoggedDist: last.cp.DistanceFromStart,
    lastLoggedAt: new Date(last.log.RecordedAt),
  }
}

// Pace between each pair of consecutive logged checkpoints (both with a known
// distance) for a runner — a full splits view, unlike computeRunnerPace, which
// only reports the most recent split. Keyed by checkpoint ID; the first
// logged-with-distance checkpoint has no entry (no prior split to compare
// against), and checkpoints that are unlogged or lack a distance never appear
// as keys — render as blank/— in the UI.
export function computeSplitPaces(
  runner: Runner,
  checkpoints: Checkpoint[],
  logs: CheckpointLog[],
): Map<number, number | null> {
  const splits = new Map<number, number | null>()
  if (hasNoMeaningfulPace(runner)) return splits

  const withDist = loggedWithDistance(runner.ID, checkpoints, logs)
  for (let i = 1; i < withDist.length; i++) {
    splits.set(withDist[i].cp.ID, pairPace(withDist[i - 1], withDist[i]))
  }
  return splits
}

// Projects arrival at a target checkpoint using a computed pace.
// Returns null if pace or target distance is unavailable.
export function projectArrival(pace: RunnerPace, targetDist: number): Date | null {
  if (pace.paceMinPerMile === null || pace.lastLoggedDist === null || pace.lastLoggedAt === null)
    return null

  const distToGo = targetDist - pace.lastLoggedDist
  if (distToGo <= 0) return null

  return new Date(pace.lastLoggedAt.getTime() + distToGo * pace.paceMinPerMile * 60000)
}

// Formats pace as "MM:SS /mi" (e.g. "12:30 /mi").
export function formatPace(paceMinPerMile: number): string {
  const mins = Math.floor(paceMinPerMile)
  const secs = Math.round((paceMinPerMile - mins) * 60)
  return `${mins}:${String(secs).padStart(2, '0')} /mi`
}

File: backend/internal/domain/pace/pace.go
// Package pace computes a runner's current pace and projected arrival at a
// future checkpoint from their logged checkpoint history. It mirrors
// frontend/src/domain/pace.ts so the two stay behaviorally identical.
package pace

import (
	"fmt"
	"math"
	"sort"
	"time"

	"github.com/kevinball/ares-bib-logger/backend/internal/domain/entity"
)

// RunnerPace holds a computed pace and the checkpoint it was derived from.
// All fields are nil when pace cannot be computed.
type RunnerPace struct {
	PaceMinPerMile *float64
	LastLoggedDist *float64
	LastLoggedAt   *time.Time
}

type loggedCheckpoint struct {
	log entity.CheckpointLog
	cp  entity.Checkpoint
}

// ComputeRunnerPace computes pace from the last two logged checkpoints (in
// display order) that both have a known distance. DNS/DNF/MOVED/FINISHED
// runners always return an empty RunnerPace — once a runner has stopped,
// velocity isn't meaningful. Use LastLoggedCheckpoint if you need their last
// known location regardless of status.
func ComputeRunnerPace(runner entity.Runner, checkpoints []entity.Checkpoint, logs []entity.CheckpointLog) RunnerPace {
	switch runner.Status {
	case entity.StatusDNS, entity.StatusDNF, entity.StatusMoved, entity.StatusFinished:
		return RunnerPace{}
	}

	loggedWithCP := loggedCheckpointsForRunner(checkpoints, logs, runner.ID)

	var withDist []loggedCheckpoint
	for _, x := range loggedWithCP {
		if x.cp.DistanceFromStart != nil {
			withDist = append(withDist, x)
		}
	}

	if len(withDist) < 2 {
		if len(withDist) == 0 {
			return RunnerPace{}
		}
		last := withDist[0]
		dist := *last.cp.DistanceFromStart
		at := last.log.RecordedAt
		return RunnerPace{LastLoggedDist: &dist, LastLoggedAt: &at}
	}

	prev := withDist[len(withDist)-2]
	last := withDist[len(withDist)-1]

	distDelta := *last.cp.DistanceFromStart - *prev.cp.DistanceFromStart
	timeDelta := last.log.RecordedAt.Sub(prev.log.RecordedAt)

	dist := *last.cp.DistanceFromStart
	at := last.log.RecordedAt

	if distDelta <= 0 || timeDelta <= 0 {
		return RunnerPace{LastLoggedDist: &dist, LastLoggedAt: &at}
	}

	paceMinPerMile := timeDelta.Minutes() / distDelta
	return RunnerPace{PaceMinPerMile: &paceMinPerMile, LastLoggedDist: &dist, LastLoggedAt: &at}
}

// ProjectArrival projects arrival at a target checkpoint using a computed
// pace. Returns nil if pace or target distance is unavailable.
func ProjectArrival(p RunnerPace, targetDist float64) *time.Time {
	if p.PaceMinPerMile == nil || p.LastLoggedDist == nil || p.LastLoggedAt == nil {
		return nil
	}
	distToGo := targetDist - *p.LastLoggedDist
	if distToGo <= 0 {
		return nil
	}
	t := p.LastLoggedAt.Add(time.Duration(distToGo * *p.PaceMinPerMile * float64(time.Minute)))
	return &t
}

// FormatPace formats pace as "MM:SS /mi".
func FormatPace(paceMinPerMile float64) string {
	mins := int(paceMinPerMile)
	secs := int(math.Round((paceMinPerMile - float64(mins)) * 60))
	return fmt.Sprintf("%d:%02d /mi", mins, secs)
}

// LastLoggedCheckpoint returns the runner's most-recently-passed checkpoint
// (max DisplayOrder among their logs), independent of status. Unlike
// ComputeRunnerPace, this never comes back empty just because the runner's
// status is terminal — a dropped runner's last known station is exactly the
// information a search-and-rescue query needs.
func LastLoggedCheckpoint(checkpoints []entity.Checkpoint, logs []entity.CheckpointLog, runnerID int) (entity.Checkpoint, entity.CheckpointLog, bool) {
	loggedWithCP := loggedCheckpointsForRunner(checkpoints, logs, runnerID)
	if len(loggedWithCP) == 0 {
		return entity.Checkpoint{}, entity.CheckpointLog{}, false
	}
	best := loggedWithCP[len(loggedWithCP)-1]
	return best.cp, best.log, true
}

// loggedCheckpointsForRunner returns this runner's logged checkpoints,
// sorted ascending by display order.
func loggedCheckpointsForRunner(checkpoints []entity.Checkpoint, logs []entity.CheckpointLog, runnerID int) []loggedCheckpoint {
	cpByID := make(map[int]entity.Checkpoint, len(checkpoints))
	for _, cp := range checkpoints {
		cpByID[cp.ID] = cp
	}

	var out []loggedCheckpoint
	for _, l := range logs {
		if l.RunnerID != runnerID {
			continue
		}
		cp, ok := cpByID[l.CheckpointID]
		if !ok {
			continue
		}
		out = append(out, loggedCheckpoint{log: l, cp: cp})
	}
	sort.Slice(out, func(i, j int) bool {
		return out[i].cp.DisplayOrder < out[j].cp.DisplayOrder
	})
	return out
}
